tensorflow/core/kernels/data/experimental/sampling_dataset_op.cc

```diff
@@ -23,202 +25,207 @@ limitations under the License.
-            writer->WriteScalar(this->full_name("seed2"), seed2_));
-
-        if (input_impl_) {
-          TF_RETURN_IF_ERROR(SaveInput(writer, input_impl_));
-        } else {
-          TF_RETURN_IF_ERROR(
-              writer->WriteScalar(full_name("input_impl_empty"), ""));
-        }
-        return Status::OK();
+            writer->WriteScalar(full_name("input_impl_empty"), ""));
       }
+      return Status::OK();
+    }
 
-      Status RestoreInternal(IteratorContext* ctx,
-                             IteratorStateReader* reader) override {
-        mutex_lock l(mu_);
-        // Restore the random number generators.
-        TF_RETURN_IF_ERROR(reader->ReadScalar(
-            this->full_name("num_random_samples"), &num_random_samples_));
-        TF_RETURN_IF_ERROR(reader->ReadScalar(this->full_name("seed"), &seed_));
-        TF_RETURN_IF_ERROR(
-            reader->ReadScalar(this->full_name("seed2"), &seed2_));
-        ResetRngs();
-
-        if (!reader->Contains(full_name("input_impl_empty"))) {
-          TF_RETURN_IF_ERROR(RestoreInput(ctx, reader, input_impl_));
-        } else {
-          input_impl_.reset();
-        }
-        return Status::OK();
+    Status RestoreInternal(IteratorContext* ctx,
+                           IteratorStateReader* reader) override {
+      mutex_lock l(mu_);
+      // Restore the random number generators.
+      TF_RETURN_IF_ERROR(reader->ReadScalar(
+          this->full_name("num_random_samples"), &num_random_samples_));
+      TF_RETURN_IF_ERROR(reader->ReadScalar(this->full_name("seed"), &seed_));
+      TF_RETURN_IF_ERROR(reader->ReadScalar(this->full_name("seed2"), &seed2_));
+      ResetRngs();
+
+      if (!reader->Contains(full_name("input_impl_empty"))) {
+        TF_RETURN_IF_ERROR(RestoreInput(ctx, reader, input_impl_));
+      } else {
+        input_impl_.reset();
       }
+      return Status::OK();
+    }
 
-      mutex mu_;
-      int64 seed_ GUARDED_BY(mu_);
-      int64 seed2_ GUARDED_BY(mu_);
-
-     private:
-      std::unique_ptr<IteratorBase> input_impl_ GUARDED_BY(mu_);
+    mutex mu_;
+    int64 seed_ GUARDED_BY(mu_);
+    int64 seed2_ GUARDED_BY(mu_);
 
-      float Random() {
-        mutex_lock l(mu_);
-        num_random_samples_++;
-        auto out = generator_.RandFloat();
-        return out;
-      }
+   private:
+    std::unique_ptr<IteratorBase> input_impl_ GUARDED_BY(mu_);
 
-      // random util
-      random::PhiloxRandom parent_generator_ GUARDED_BY(mu_);
-      random::SimplePhilox generator_ GUARDED_BY(mu_);
-      int64 num_random_samples_ GUARDED_BY(mu_) = 0;
-    };
+    float Random() {
+      mutex_lock l(mu_);
+      num_random_samples_++;
+      auto out = generator_.RandFloat();
+      return out;
+    }
 
-    const float rate_;
-    const int64 seed_, seed2_;
-    const DatasetBase* const input_;
+    // random util
+    random::PhiloxRandom parent_generator_ GUARDED_BY(mu_);
+    random::SimplePhilox generator_ GUARDED_BY(mu_);
```

Comment:
If you make this `random::SingleSampleAdapter<random::PhiloxRandom>` similar to what `shuffle_dataset_op.cc` does, you will not need to change the API of `random::SimplePhilox` which would be good to avoid.

Reply:
I've pushed some new changes in commit f1406b4 that make `SampleDataset` use `random::SingleSampleAdapter<random::PhiloxRandom>`.